src/libcapsicum_host.3
.\"
.\" Copyright (c) 2009 Robert N. M. Watson
.\" All rights reserved.
.\"
.\" WARNING: THIS IS EXPERIMENTAL SECURITY SOFTWARE THAT MUST NOT BE RELIED
.\" ON IN PRODUCTION SYSTEMS.  IT WILL BREAK YOUR SOFTWARE IN NEW AND
.\" UNEXPECTED WAYS.
.\"
.\" This software was developed at the University of Cambridge Computer
.\" Laboratory with support from a grant from Google, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHORS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd June 11, 2009
.Os
.Dt LIBCAPSICUM_HOST 3
.Sh NAME
.Nm libcapsicum
.Nd "library interface to capability-mode services"
.Sh LIBRARY
.Lb libcapsicum
.Sh SYNOPSIS
.In sys/types.h
.In sys/capability.h
.In libcapsicum.h
.Ft int
.Fn lch_start "const char *sandbox" "char *const argv[]" "u_int flags" "struct lc_sandbox **lcsp"
.Ft int
.Fn lch_startfd "int fd_sandbox" "char *const argv[]" "u_int flags" "struct lc_sandbox **lcsp"
.Ft void
.Fn lch_stop "struct lc_sandbox *lcsp"
.Ft int
.Fn lch_autosandbox_isenabled "const char *servicename"
.Ft int
.Fn lch_getsock "struct lc_sandbox *lcsp" "int *fdp"
.Ft int
.Fn lch_getpid "struct lc_sandbox *lcsp" "pid_t *pidp"
.Ft int
.Fn lch_getprocdesc "struct lc_sandbox *lcsp" "int *fdp"
.Ft ssize_t
.Fn lch_recv "struct lc_sandbox *lcsp, void *buf" "size_t len" "int flags"
.Ft ssize_t
.Fn lch_recv_rights "struct lc_sandbox *lcsp" "void *buf" "size_t len" "int flags" "int *fdp" "int *fdcountp"
.Ft int
.Fn lch_rpc "struct lc_sandbox *lcsp" "u_int32_t opno" "struct iovec *req" "int reqcount" "struct iovec *rep" "int repcount" "size_t *replenp"
.Ft int
.Fn lch_rpc_rights "struct lc_sandbox *lcsp" "u_int32_t opno" "struct iovec *req" "int reqcount" "int *req_fdp" "int req_fdcount" "struct iovec *rep" "int repcount" "size_t *replenp" "int *rep_fdp" "int *rep_fdcountp"
.Ft ssize_t
.Fn lch_send "struct lc_sandbox *lcsp" "const void *msg" "size_t len" "int flags"
.Ft ssize_t
.Fn lch_send_rights "struct lc_sandbox *lcsp" "const void *msg" "size_t len" "int flags" "int *fdp" "int fdcount"
.Sh DESCRIPTION
The
.Nm
library routines provide services for processes hosting or running in
capability mode.
Depending on the requirements of the host and sandbox, the API can simply be
used to set up and stop sandboxes, used to manage I/O using a
.Xr unix 4
domain socket connection to the sandbox, or can provide a basic remote
procedure call (RPC) facility.
Applications may also use RPC generators such as
.Xr rpcgen 1
to build event handling and marshaling code.
.Pp
This man page describes the host API.
General information on
.Nm
may be found in
.Xr libcapsicum 3 .
Information on the sandbox API may be found in
.Xr libcapsicum_sandbox 3 .
.Sh HOST API
The
.Nm
host API allows processes to start, stop, and manage sandboxes running in
capability mode.
Host API functions can be identified by their function name prefix,
.Dv lch_ .
.Pp
Each executing sandbox instance is described by an opaque
.Dt "struct lc_sandbox *" ,
which is returned by
.Fn lch_start
for successfully started sandboxes, and passed into other APIs to indicate
which sandbox should be acted on.
.Fn lch_start
creates a new executing sandboxes, given the name of the sandbox binary via
.Va sandbox ,
and command line arguments
.Va argv ,
and optional flags
.Va flags
to fine-tune aspects of sandbox operation; the only currently defined flag is
.Dv LCH_PERMIT_STDERR ,
which allows the sandbox to write to the current process's
.Dv stderr .
By default, this is not permitted.
.Pp
.Fn lch_startfd
accept a file descriptor argument,
.Va fd_sandbox ,
rather than a path, so is appropriate for use within a sandbox.
.Pp
Executing sandboxes may be stopped (and all state freed) using
.Fn lch_stop .
Following a call to
.Fn lch_stop ,
the
.Va lchp
argument will no longer be valid.
.Pp
Libraries and tools performing self-compartmentalization can use the
interface
.Nm lch_autosandbox_isenabled
along with a unique string identifying their service to determine whether or
not a global policy affecting the service requires sandboxing to be enabled
or not.
.Pp
Properties of the sandbox, such as the socket used to communicate with it,
the proces descriptor for the sandbox process, and the pid, may be queried
using
.Fn lch_getsock ,
.Fn lch_getprocdesc ,
and
.Fn lch_getpid .
.Pp
.Nm
implements a number of I/O functions as part of the host API, which are
documented in
.Xr libcapsicum_host 3 .
.Fn lch_recv
and
.Fn lch_send
provide simple wrappers around
.Xr recv 2
and
.Xr send 2
to avoid sandbox consumers from having to query sandbox socket file
descriptors before use.
.Pp
.Fn lch_recv_rights
and
.Fn lch_send_rights
are similar, but allow file descriptors to be attached the the messages
received and sent.
Both accept a pointer to a file descriptor array,
.Va fdp .
Callers to
.Fn lch_recv_rights
will pass in the length of the array via
.Va fdcountp ,
whose value will be changed to the actual number of file descriptors
received.
Callers to
.Fn lch_send_rights
will pass in the number of file descriptors in the array via
.Va fdcount .
.Pp
.Fn lch_rpc
provides a simple synchronous RPC facility, and is intended to be used in
coordination with the
.Fn lcs_recvrpc
and
.Fn lcs_sendrpc
sandbox APIs.
The host provides an operation number meaningful to the sandbox,
.Va opno,
RPC arguments represented by
.Va req
and
.Va reqcount
using an
.Vt iovec
in the style of
.Xr writev 2 ,
and similar receive buffers passed via
.Va rep
and
.Va repcount .
If the RPC fails, -1 will be returned, or 0 and the size of any reply will be
returned by reference using
.Va replenp .
.Nm lch_rpc_rights
allows the sending and receiving of file descriptors as part of the RPC
operation.
.Sh SEE ALSO
.Xr rpcgen 1 ,
.Xr recv 2 ,
.Xr send 2 ,
.Xr writev 2 ,
.Xr free 3 ,
.Xr libcapsicum 3 ,
.Xr libcapsicum_sandbox 3 ,
.Xr malloc 3 ,
.Xr unix 4
.Sh HISTORY
Support for capabilities and capabilities mode was developed as part of the
.Tn TrustedBSD
Project.
.Sh BUGS
WARNING: THIS IS EXPERIMENTAL SECURITY SOFTWARE THAT MUST NOT BE RELIED ON IN
PRODUCTION SYSTEMS.  IT WILL BREAK YOUR SOFTWARE IN NEW AND UNEXPECTED WAYS.
.Pp
All sequence numbers will always have the value 0.
This is fine from a retransmission perspective, as generally no
retransmission should be required, but consumers should serialize use of the
RPC service when consuming it from concurrent callers (such as multiple
threads or multiple processes) to prevent I/O interlacing from corrupting the
RPC stream.
.Sh AUTHORS
These functions and the capability facility were created by
.An "Robert N. M. Watson"
at the University of Cambridge Computer Laboratory with support from a grant
from Google, Inc.


src/libcapsicum_host.c
/*-
 * Copyright (c) 2009-2010 Robert N. M. Watson
 * All rights reserved.
 *
 * This software was developed at the University of Cambridge Computer
 * Laboratory with support from a grant from Google, Inc.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/param.h>
#include <sys/capability.h>
#include <sys/mman.h>
#include <sys/procdesc.h>
#include <sys/sbuf.h>
#include <sys/socket.h>
#include <sys/uio.h>

#define _WITH_DPRINTF
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <libgen.h>
#include <paths.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "libcapsicum.h"
#include "libcapsicum_internal.h"
#include "libcapsicum_sandbox_api.h"

#define	LIBCAPSICUM_CAPMASK_SOCK 					\
	(CAP_POLL_EVENT | CAP_READ | CAP_WRITE)

#define	LIBCAPSICUM_CAPMASK_BIN	\
	(CAP_READ | CAP_POLL_EVENT | CAP_FSTAT | CAP_FSTATFS |		\
	 CAP_FEXECVE | CAP_MMAP | CAP_MAPEXEC)

#define LIBCAPSICUM_CAPMASK_LIBDIR					\
	LIBCAPSICUM_CAPMASK_BIN | CAP_LOOKUP

#define LIBCAPSICUM_CAPMASK_FDLIST					\
	CAP_READ | CAP_WRITE | CAP_FTRUNCATE | CAP_FSTAT | CAP_MMAP

#define LIBCAPSICUM_CAPMASK_STDOUT					\
	CAP_WRITE | CAP_SEEK | CAP_FSTAT

extern char **environ;

#define LD_ELF_CAP_SO		"ld-elf-cap.so.1"
#define	PATH_LD_ELF_CAP_SO	"/libexec"

int
lch_autosandbox_isenabled(__unused const char *servicename)
{

	if (getenv("LIBCAPSICUM_NOAUTOSANDBOX") != NULL)
		return (0);
	return (1);
}

/*
 * Once in the child process, create the new sandbox.
 *
 * XXX: A number of things happen here that are not safe after fork(),
 * especially calls to err().
 */
static void
lch_sandbox(int fd_sock, int fd_binary, int fd_rtld, u_int flags,
    const char *binname, char *const argv[], struct lc_fdlist *userfds)
{
	struct sbuf *sbufp;
	int shmfd = -1;
	struct lc_fdlist *fds;

	/*
	 * Inform the run-time linked of the binary's name.
	 */
	if (setenv("LD_BINNAME", binname, 1) == -1)
		err(-1, "Error in setenv(LD_BINNAME)");

	/*
	 * Create an anonymous shared memory segment for the FD list.
	 */
	shmfd = shm_open(SHM_ANON, O_RDWR | O_CREAT | O_TRUNC, 0600);
	if (shmfd < 0)
		err(-1, "Error creating shared memory segment");

	/*
	 * Create and fill up the FD list.
	 */
	fds = lc_fdlist_new();
	if (fds == NULL)
		err(-1, "Error in lc_fdlist_new()");

	if (lc_fdlist_addcap(fds, LIBCAPSICUM_FQNAME, "stdin", "",
		STDIN_FILENO, 0) < 0)
		err(-1, "Error in lc_fdlist_addcap(stdin)");

	if (lc_fdlist_addcap(fds, LIBCAPSICUM_FQNAME, "stdout", "",
		STDOUT_FILENO,
		(flags & LCH_PERMIT_STDOUT) ? LIBCAPSICUM_CAPMASK_STDOUT : 0) < 0)
		err(-1, "Error in lc_fdlist_addcap(stdout)");

	if (lc_fdlist_addcap(fds, LIBCAPSICUM_FQNAME, "stderr", "",
		STDERR_FILENO,
		(flags & LCH_PERMIT_STDERR) ? LIBCAPSICUM_CAPMASK_STDOUT : 0) < 0)
		err(-1, "Error in lc_fdlist_addcap(stderr)");

	if (lc_fdlist_addcap(fds, LIBCAPSICUM_FQNAME, "socket", "",
	                     fd_sock, LIBCAPSICUM_CAPMASK_SOCK) < 0)
		err(-1, "Error in lc_fdlist_addcap(fd_sock)");

	if (lc_fdlist_addcap(fds, LIBCAPSICUM_FQNAME, "fdlist", "",
	                     shmfd, LIBCAPSICUM_CAPMASK_FDLIST) < 0)
		err(-1, "Error in lc_fdlist_addcap(shmfd)");

	if (lc_fdlist_addcap(fds, RTLD_CAP_FQNAME, "rtld", "",
	                     fd_rtld, LIBCAPSICUM_CAPMASK_BIN) < 0)
		err(-1, "Error in lc_fdlist_addcap(fd_rtld)");

	if (lc_fdlist_addcap(fds, RTLD_CAP_FQNAME, "Executable", binname,
	                     fd_binary, LIBCAPSICUM_CAPMASK_BIN) < 0)
		err(-1, "Error in lc_fdlist_addcap(fd_binary)");

	/*
	 * Ask RTLD for library path descriptors.
	 *
	 * NOTE: This is FreeBSD-specific; porting to other operating systems
	 * will require dynamic linkers capable of answering similar queries.
	 */
	int size = 16;
	int *libdirs;

	while (1) {
		libdirs = malloc(size * sizeof(int));
		if (ld_libdirs(libdirs, &size) < 0) {
			free(libdirs);
			if (size > 0)
				continue;
			err(-1, "Error in ld_libdirs()");
		} else
			break;
	}

	for (int j = 0; j < size; j++) {
		if (lc_fdlist_addcap(fds, RTLD_CAP_FQNAME, "LibraryDirectory", "",
		    libdirs[j], LIBCAPSICUM_CAPMASK_LIBDIR) < 0)
			err(-1, "Error in lc_fdlist_addcap(libdirs[%d]: %d)",
			    j, libdirs[j]);
	}

	/* Append user FD list and reorder the descriptors */
	if (lc_fdlist_append(fds, userfds) < 0)
		err(-1, "Error in lc_fdlist_append()");

	if (lc_fdlist_reorder(fds) < 0)
		err(-1, "Error in lc_fdlist_reorder()");




	/*
	 * Find the fdlist shared memory segment.
	 */
	int pos = 0;
	if (lc_fdlist_lookup(fds, LIBCAPSICUM_FQNAME, "fdlist", NULL, &shmfd,
	    &pos) < 0)
		err(-1, "Error in lc_fdlist_lookup(fdlist)");

	if (lc_fdlist_set_global(fds, shmfd) != 0)
		err(-1, "Error in lc_fdlist_set_global()");

	/*
	 * Find RTLD.
	 */
	if (lc_fdlist_lookup(fds, RTLD_CAP_FQNAME, "rtld", NULL, &fd_rtld,
	                     NULL) < 0)
		err(-1, "Error in lc_fdlist_lookup(RTLD)");

	/*
	 * Find the binary for RTLD.
	 */
	if (lc_fdlist_lookup(fds, RTLD_CAP_FQNAME, "Executable", NULL,
	    &fd_binary, NULL) < 0)
		err(-1, "Error in lc_fdlist_lookup(Executable)");

	char tmp[8];
	sprintf(tmp, "%d", fd_binary);
	if (setenv("LD_BINARY", tmp, 1) != 0)
		err(-1, "Error in setenv(LD_BINARY)");

	/*
	 * Build LD_LIBRARY_DIRS and LD_PRELOAD for RTLD.
	 *
	 * NOTE: This is FreeBSD-specific; porting to other operating systems
	 * will require dynamic linkers capable of operating on file
	 * descriptors.
	 */
	sbufp = sbuf_new_auto();
	if (sbufp == NULL)
		err(-1, "Error in sbuf_new_auto()");

	{
		int fd;
		pos = 0;
		while (lc_fdlist_lookup(fds, RTLD_CAP_FQNAME, "LibraryDirectory",
		       NULL, &fd, &pos) >= 0)
			sbuf_printf(sbufp, "%d:", fd);
	}

	sbuf_finish(sbufp);
	if (setenv("LD_LIBRARY_DIRS", sbuf_data(sbufp), 1) == -1)
		err(-1, "Error in setenv(LD_LIBRARY_DIRS)");
	sbuf_delete(sbufp);

	sbufp = sbuf_new_auto();
	if (sbufp == NULL)
		err(-1, "Error in sbuf_new_auto()");

	{
		int fd;
		pos = 0;
		while (lc_fdlist_lookup(fds, RTLD_CAP_FQNAME, "PreloadLibrary",
		       NULL, &fd, &pos) >= 0)
			sbuf_printf(sbufp, "%d:", fd);
	}

	sbuf_finish(sbufp);
	if (setenv("LD_PRELOAD", sbuf_data(sbufp), 1) == -1)
		err(-1, "Error in setenv(LD_PRELOAD)");
	sbuf_delete(sbufp);

	if (cap_enter() < 0)
		err(-1, "cap_enter() failed");

	(void)fexecve(fd_rtld, argv, environ);
	dprintf(2, "ERROR: fexecve() failed; errno = %d\n", errno);
}

int
lch_startfd(int fd_binary, const char *binname, char *const argv[],
    u_int flags, struct lc_fdlist *fds, struct lc_sandbox **lcspp)
{
	struct lc_sandbox *lcsp;
	int fd_rtld;
	int fd_procdesc, fd_sockpair[2];
	int error, val;
	pid_t pid;

	fd_rtld = fd_procdesc = fd_sockpair[0] = fd_sockpair[1] = -1;

	lcsp = malloc(sizeof(*lcsp));
	if (lcsp == NULL)
		return (-1);
	bzero(lcsp, sizeof(*lcsp));

	if (ld_insandbox()) {
		struct lc_fdlist *globals;
		int pos = 0;

		globals = lc_fdlist_global();
		if (globals == NULL)
			goto out_error;
		if (lc_fdlist_lookup(globals, RTLD_CAP_FQNAME, "rtld", NULL,
		    &fd_rtld, &pos) < 0)
			goto out_error;
	} else {
		fd_rtld = open(PATH_LD_ELF_CAP_SO "/" LD_ELF_CAP_SO,
		    O_RDONLY);
		if (fd_rtld < 0)
			goto out_error;
	}

	if (socketpair(PF_LOCAL, SOCK_STREAM, 0, fd_sockpair) < 0)
		goto out_error;

	val = 1;
	if (setsockopt(fd_sockpair[0], SOL_SOCKET, SO_NOSIGPIPE, &val,
	    sizeof(val)) < 0) {
		fd_sockpair[0] = fd_sockpair[1] = -1;
		goto out_error;
	}

	pid = pdfork(&fd_procdesc);
	if (pid < 0) {
		fd_procdesc = -1;
		goto out_error;
	}
	if (pid == 0) {
		lch_sandbox(fd_sockpair[1], fd_binary, fd_rtld, flags,
		    binname, argv, fds);
		exit(-1);
	}
	if (fd_rtld != -1)
		close(fd_rtld);
	close(fd_sockpair[1]);

	lcsp->lcs_fd_procdesc = fd_procdesc;
	lcsp->lcs_fd_sock = fd_sockpair[0];
	lcsp->lcs_pid = pid;
	*lcspp = lcsp;

	return (0);

out_error:
	error = errno;
	if (fd_sockpair[0] != -1)
		close(fd_sockpair[0]);
	if (fd_sockpair[1] != -1)
		close(fd_sockpair[1]);
	if (fd_rtld != -1)
		close(fd_rtld);
	if (lcsp != NULL)
		free(lcsp);
	errno = error;
	return (-1);
}

int
lch_start(const char *sandbox, char *const argv[], u_int flags,
    struct lc_fdlist *fds, struct lc_sandbox **lcspp)
{
	char binname[MAXPATHLEN];
	int error, fd_binary, ret;

	if (basename_r(sandbox, binname) == NULL)
		return (-1);

	fd_binary = open(sandbox, O_RDONLY);
	if (fd_binary < 0)
		return (-1);

	ret = lch_startfd(fd_binary, binname, argv, flags, fds, lcspp);
	error = errno;
	close(fd_binary);
	errno = error;
	return (ret);
}

void
lch_stop(struct lc_sandbox *lcsp)
{

	close(lcsp->lcs_fd_sock);
	close(lcsp->lcs_fd_procdesc);
	lcsp->lcs_fd_sock = -1;
	lcsp->lcs_fd_procdesc = -1;
	lcsp->lcs_pid = -1;
	free(lcsp);
}

int
lch_getsock(struct lc_sandbox *lcsp, int *fdp)
{

	*fdp = lcsp->lcs_fd_sock;
	return (0);
}

int
lch_getpid(struct lc_sandbox *lcsp, pid_t *pidp)
{

	*pidp = lcsp->lcs_pid;
	return (0);
}

int
lch_getprocdesc(struct lc_sandbox *lcsp, int *fdp)
{

	*fdp = lcsp->lcs_fd_procdesc;
	return (0);
}
